src/projection.jl

```diff
@@ -0,0 +1,77 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    projector([T::Type], x)
+
+Returns a `project(dx)` closure which maps `dx` onto type `T`, such that it is the
+same size as `x`. If `T` is not provided, it is assumed to be the type of `x`.
+
+It's necessary to have `x` to ensure that it's possible to project e.g. `AbstractZero`s
+onto `Array`s -- this wouldn't be possible with type information alone because the neither
+`AbstractZero`s nor `T` know what size of `Array` to produce.
+"""
+function projector end
+
+projector(x) = projector(typeof(x), x)
+
+# fallback (structs)
+function projector(::Type{T}, x::T) where T
+    project(dx::T) = dx
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Numbers
+function projector(::Type{T}, x::T) where {T<:Real}
+    project(dx::Real) = T(dx)
+    project(dx::Number) = T(real(dx)) # to avoid InexactError
```

Comment:
I think this is too tight, as `projector(2)(3.5)` is going to be an InexactError right? As is `projector(false)(1.5)`.

And more generally, what if (say) I want to put dual numbers into the pullback? My impression is that that should be allowed. Which is what led me to think that only known problems should be projected out, like `dx::Complex` when `x::Real`, or anything when `x::Bool`. But it would be nice if the door were open for packages to add to the list of "things which get projected like Complex -> Real".

Reply:
Yeah that's a relatively serious downside

Reply:
Seems to have made it into the tagged version:
```
julia> ProjectTo(1)(2.5)
ERROR: InexactError: Int64(2.5)

(jl_5kFIPa) pkg> st ChainRulesCore
      Status `/private/var/folders/yq/4p2zwd614y59gszh7y9ypyhh0000gn/T/jl_5kFIPa/Project.toml`
  [d360d2e6] ChainRulesCore v0.10.11
```